### ISSUE TITLE
[Bug]: AC0032 - False Postive when FindSet is inside a procedure (of diff object)

### ISSUE THREAD

**Issue by pri-kise** [ALCops/Analyzers#466]
### Analyzer

ApplicationCop

### Issue type

False positive (diagnostic fires when it shouldn't)

### Rule ID

AC0032

### Description

There's a wrong warning when you have a procedure that reads the data and call this instead of having a findset inside your object.

### AL code to reproduce

```al
codeunit 50100 "PTE Cust. Integr."
{
    Access = Internal;
    InherentPermissions = X;
    Permissions =
        tabledata "PTE Doc. Entry" = r;

    internal procedure UpdateCustDocs(SalesCrMemoHeader: Record "Sales Cr.Memo Header")
    var        
        DocEntry: Record "PTE Doc. Entry";
        UpdateHelper: Codeunit "PTE Upd. Helper";
    begin
      
            if UpdateHelper.FindDocEntries(SalesCrMemoHeader.RecordId(),  DocEntry) then
                repeat
                        //UpdateHelper.DoMagic2(DocEntry);
                until DocEntry.Next() = 0;
    end;
}
```
```al
codeunit 50177 "PTE Upd. Helper"
{
    Access = Internal;
    InherentPermissions = X;
    Permissions = Tabledata "PTE Doc. Entry" = R;


    internal procedure FindDocEntries(RecId: RecordId; var DocEntry: Record "PTE Doc. Entry"): Boolean
    begin
        DocEntry.Reset();
        DocEntry.SetRange("Record ID", RecId);
        exit(DocEntry.FindSet());
    end;
}
```

**Comment by Arthurvdv** [comment]
## Investigation & recommendation

Thanks for the clear repro. Confirmed — and it's not really a cross-object problem, it's a missing read operator.

### Root cause

Both AC0031 and AC0032 decide "does this call touch the database?" through a single shared table, `MethodOperationMap` (`src/ALCops.Common/Permissions/MethodOperationMap.cs`). It maps `Find`, `FindFirst`, `FindLast`, `FindSet`, `Get`, `GetBySystemId`, `IsEmpty`, `Count` → Read, plus the I/M/D methods. **`Next` is not in the map.**

In the repro the only mapped call on `"PTE Doc. Entry"` inside `PTE Cust. Integr.` would be `DocEntry.Next()` in the `until` clause, and that is invisible to the analyzer. `FindSet` lives in `PTE Upd. Helper`, so from AC0032's point of view the codeunit never reads the table and `tabledata "PTE Doc. Entry" = r` is reported as unused.

The declared permission is genuinely required at runtime: `Next()` advances the server-side cursor and fetches the next row (or next page of rows), which is a read against the table performed *by the calling object*. Permissions don't flow through the call stack, so `PTE Cust. Integr.` needs `r` for itself. AC0032 is therefore wrong here, and — by the same reasoning — AC0031 currently fails to *demand* `r` in this pattern (a caller iterating a set filled by another object with no `Permissions` entry gets no diagnostic today).

### Why not trace into `UpdateHelper.FindDocEntries`?

Cross-object tracing was considered and rejected: the `Permissions` property is per-object and is not inherited through calls (see Known limitation 4 in the AC0032 instructions). Following the `var` parameter into another codeunit would answer the wrong question. Recognizing `Next` as a read operator on the caller's own record variable is the correct and much smaller fix.

### Supporting evidence from the SDK / Microsoft cops

- `Next` is a built-in on both record and RecordRef types:
  `Microsoft.Dynamics.Nav.CodeAnalysis/.../Symbols/TableClassTypeSymbol.cs:28` and `.../Symbols/RecordRefClassTypeSymbol.cs:50` (`BuiltInMethodTypeSymbol(..., "Next", ...)`, returns Integer, optional `Steps` parameter).
- Microsoft's own CodeCop treats `Next` as a first-class data-set invocation: `Microsoft.Dynamics.Nav.CodeCop/.../CodeCop.Design/RecDbInvocationAnalyzer.cs` has `InvocationType.Next` next to `FindDataSet`/`FindSingle` (AA0181/AA0233). Notably line ~79 special-cases *exactly this scenario*: a `Next` with no preceding `FindSet` is allowed when the record `symbol.Kind == SymbolKind.Parameter` (the set was positioned by the caller/callee via a `var` record). Microsoft accepts this as a legitimate iteration pattern, so we should count its DB read.
- `Microsoft.Dynamics.Nav.CodeCop/.../CodeCop.Utilities/BuiltInMethodNames.cs:79` — `Next = "next"`.

### Recommended fix (small)

1. **`MethodOperationMap`** — add one entry under read operations:
   ```csharp
   new KeyValuePair<string, DatabaseOperation>("Next", DatabaseOperation.Read),
   ```
   Everything downstream picks it up automatically:
   - AC0031: `RegisterOperationAction` → `RequiredPermissionDetector.TryGetFromInvocation` → `MethodOperationMap.GetOperation`.
   - AC0032: `HasPossibleDbInvocation` pre-filter, `TryGetPermissionFromDbAccess` (both `Next()` and paren-less `Next`), `TryGetPermissionViaSymbolInfo` fallback and the RecordRef whole-object bailout all consult `MethodOperationMap.GetOperation`.
   No syntax-shape analysis (repeat/until) is needed or wanted: a stand-alone `Rec.Next()` / `Rec.Next(-1)` / `while Rec.Next() <> 0` all read the DB.

2. **Do *not* add `Next` to `RecordMethodClassification.ReadMethods` / `SingleRecordReadMethods`** (`src/ALCops.Common/RecordMethodClassification.cs`). Those lists feed AC0030 (`UseReturnValueForDatabaseReadMethods`) and PC `PartialRecordOperations`, where `Next` would be wrong (its return value is legitimately ignored in `until ... = 0` and it does not load a buffer with partial-record semantics). Update the `ReadMethods` doc comment, which currently claims to be "derived from `MethodOperationMap`", so the intentional divergence is documented.

3. **Decisions already taken** (so the implementer doesn't need to re-open them):
   - Stand-alone `Next()` counts as Read (not only inside `repeat..until`).
   - AC0031 gets the symmetric behaviour: `Next()` on a record without `r` is reported. This is a bug fix (true positive at runtime), but it may surface *new* AC0031 diagnostics in existing projects — call this out in the release notes.
   - `RecordRef.Next()` triggers the existing AC0032 whole-object bailout, consistent with `RecordRef.FindSet()` (issue #420). No extra code.
   - Scope is `Next` only. `CalcFields`/`CalcSums` remain a known limitation.

### Test cases to add

`src/ALCops.ApplicationCop.Test/Rules/TableDataAccessUnusedPermissions/`
- `NoDiagnostic/NextOnVarParameterFilledElsewhere.al` — the repro: helper codeunit does `FindSet`, caller only does `repeat ... until DocEntry.Next() = 0` with `= r`.
- `NoDiagnostic/NextWithoutParentheses.al` — `until MyRecord.Next = 0`.
- `NoDiagnostic/NextStandAlone.al` — `MyRecord.Next(-1);` outside a loop.
- `NoDiagnostic/RecordRefNext.al` — `RecRef.Next()` on a RecordRef → bailout, no diagnostic even with extra unused entries.
- `HasDiagnostic/NextPartialUnused.al` — only `Next()` used, declared `rimd` → report `imd` unused (`UsedChars = "r"`).

`src/ALCops.ApplicationCop.Test/Rules/TableDataAccessRequiresPermissions/`
- `HasDiagnostic/NextWithoutPermissions.al` — `var` record param iterated with `Next()`, no `Permissions` → AC0031 for `r`.
- `NoDiagnostic/NextWithPermissions.al` — same with `Permissions = tabledata X = r`.
- `NoDiagnostic/PageSourceTableNext.al` — `Rec.Next()` in a page trigger stays covered by the SourceTable exemption.
- `NoDiagnostic/TemporaryRecordNext.al` — `Next()` on a temporary record requires nothing.

### Docs / instruction files to update

- `.github/instructions/ac0031-table-data-access-requires-permissions.instructions.md` (read-method list at line ~55, test-case counts).
- `.github/instructions/ac0032-table-data-access-unused-permissions.instructions.md` (add a design-decision row for `Next`, test-case counts).
- `.github/instructions/common-library.instructions.md` if it enumerates the map.
- Rule pages for AC0031/AC0032 on alcops.dev (separate docs repo): mention `Next` as a read operation and the "set positioned by another object" pattern.

**Comment by Arthurvdv** [comment]
I've published a new beta release, `v1.2.0-beta.1`, which should resolve this issue.

Could you verify for me if it's fixed on your end? Would be great to hear whether it's resolved or if you're still running into problems.

If you're using the [ALCops VS Code extension](https://marketplace.visualstudio.com/items?itemName=Arthurvdv.alcops), you can switch to the beta channel in the extension settings. To force an update check, run the `ALCops: Check for Updates` command from the command palette.

**Comment by Arthurvdv** [comment]
This should now be resolved with the release of `v1.2.0` of ALCops.

When you have a chance, could you update and verify whether the issue is fixed?

If everything is working as expected, feel free to close this issue. If you still run into the problem, please leave the issue open and share any additional context, code samples, or reproduction steps that might help me investigate further.
